src/main/java/com/google/cloud/tools/app/ProcessCaller.java

```diff
@@ -0,0 +1,189 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools.app;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.logging.Logger;
+
+/**
+ * Calls external tools like gcloud or dev_appserver. Hides system process invocation logic from
+ * {@link Action}.
+ *
+ * <p>All the logic for generating and running commands is contained in this class.
+ */
+public class ProcessCaller {
+
+  // TODO(joaomartins): Will this work in e.g., Windows?
+  private static File DEFAULT_WORKING_DIR = new File("/");
+  private static Logger LOG = Logger.getLogger(ProcessCaller.class.getName());
+  private List<String> command;
+  private File workingDirectory;
+  private boolean synchronous;
+  private String cloudSdkLocation = System.getProperty("user.home") + "/google-cloud-sdk";
+
+  private ProcessCaller(Tool tool, Collection<String> arguments) {
+    this(tool, arguments, ImmutableMap.<Option, String>of());
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory) {
+    this(tool, arguments, flags, workingDirectory, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      boolean synchronous) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, synchronous);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory, boolean synchronous) {
+    this.workingDirectory = workingDirectory;
+    this.synchronous = synchronous;
+    this.command = prepareCommand(tool, arguments, flags);
+  }
+
+  public boolean call() throws GCloudExecutionException {
+    LOG.info("Calling " + Joiner.on(" ").join(command));
+
+    ProcessBuilder processBuilder = new ProcessBuilder(command);
+    processBuilder.directory(workingDirectory);
+    processBuilder.inheritIO();
+    try {
+      Process gcloudProcess = processBuilder.start();
+
+      // If call is synchronous, wait for the process to end.
+      if (synchronous) {
+        int exitStatus = gcloudProcess.waitFor();
+
+        if (exitStatus != 0) {
+          throw new GCloudExecutionException(exitStatus);
+        }
+      }
+    } catch (IOException|InterruptedException ex) {
```

Comment:
I originally caught IOExceptions because I didn't want to add them to execute()'s throws list. But I added it in the meantime, so I guess I can get rid of it here. I can also add InterruptedException to the throws list, but I'd like to keep it short and I don't see this exception happening very much.

Do you think I should add it to execute's throws list and get rid of this try catch block?